src/dune_rules/pp_spec_rules.ml

```diff
@@ -0,0 +1,442 @@
+    ; main_module_name
+    ; loc = Loc.none
+    }
+  in
+  let obj_dir = Obj_dir.for_pp ~dir in
+  let* cctx =
+    let* expander = Super_context.expander sctx ~dir in
+    let requires_compile = Resolve.map driver_and_libs ~f:snd in
+    let requires_link = Memo.lazy_ (fun () -> Memo.return requires_compile) in
+    let flags = Ocaml_flags.of_list [ "-g"; "-w"; "-24" ] in
+    let opaque = Compilation_context.Explicit false in
+    let modules = Modules.singleton_exe module_ in
+    Compilation_context.create ~super_context:sctx ~scope ~expander ~obj_dir
+      ~modules ~flags
+      ~requires_compile:(Memo.return requires_compile)
+      ~requires_link ~opaque ~js_of_ocaml:None ~package:None ~bin_annot:false ()
+  in
+  let+ (_ : Exe.dep_graphs) =
+    Exe.build_and_link ~program ~linkages cctx ~promote:None
+  in
+  ()
+
+let get_rules sctx key =
+  let ctx = Super_context.context sctx in
+  let exe = Ppx_driver.ppx_exe ctx ~key in
+  let* pp_names, scope =
+    match Digest.from_hex key with
+    | None ->
+      User_error.raise
+        [ Pp.textf "invalid ppx key for %s"
+            (Path.Build.to_string_maybe_quoted exe)
+        ]
+    | Some key ->
+      let { Key.Decoded.pps; project_root } = Key.decode key in
+      let+ scope =
+        let dir =
+          match project_root with
+          | None -> ctx.build_dir
+          | Some dir -> Path.Build.append_source ctx.build_dir dir
+        in
+        Scope.DB.find_by_dir dir
+      in
+      (pps, scope)
+  in
+  let open Memo.O in
+  let* pps =
+    let lib_db = Scope.libs scope in
+    List.map pp_names ~f:(fun x -> (Loc.none, x)) |> Lib.DB.resolve_pps lib_db
+  in
+  build_ppx_driver sctx ~scope ~pps ~pp_names ~target:exe
+
+let gen_rules sctx components =
+  match components with
+  | [ key ] -> get_rules sctx key
+  | _ -> Memo.return ()
+
+let promote_correction fn build ~suffix =
+  let open Action_builder.O in
+  let+ act = build in
+  Action.Full.reduce
+    [ act
+    ; Action.Full.make
+        (Action.diff ~optional:true (Path.build fn)
+           (Path.Build.extend_basename fn ~suffix))
+    ]
+
+let promote_correction_with_target fn build ~suffix =
+  Action_builder.progn
+    [ build
+    ; Action_builder.with_no_targets
+        (Action_builder.return
+           (Action.Full.make
+              (Action.diff ~optional:true (Path.build fn)
+                 (Path.Build.extend_basename fn ~suffix))))
+    ]
+
+let sandbox_of_setting = function
+  | `Set_by_user d | `Default d -> d
+
+let action_for_pp ~dir ~sandbox ~loc ~expander ~action ~src ~lib_name ~ml_kind =
+  let chdir = (Expander.context expander).build_dir in
+  let expander =
+    let bindings =
+      Pform.Map.of_list_exn
+        [ (Var Input_file, [ Value.Path (Path.build src) ]) ]
+    in
+    Expander.add_bindings expander ~bindings
+  in
+  let+ action =
+    let+ action =
+      let ml_kind = Some ml_kind in
+      Action_unexpanded.expand action ~chdir ~loc ~expander ~deps:[] ~lib_name
```

Comment:
Shouldn't this be `Action_unexpanded.expand_no_targets`?

Reply:
I don't know. On line 166 below we do:

```ocaml
      Action.Full.map build ~f:(Action.with_stdout_to target))
```

Reply:
Yes, but that's different. We're adding a target ourselves, but we shouldn't allow users to add actions that introduce targets. For example, I think you're now allowing actions such as `(progn (with-stdout-to unrelated (echo foo)) ...)` which produce this `unrelated ` target.

Reply:
Weird. For normal pp, we redirect the action's stdout to the target. With ppx only, we use the `-o` argument. I would stick to using stdout.

Reply:
I think we're speaking past each other. I think your example is now valid because of the intermediate files produced by intermediate actions.

Reply:
I don't quite remember, but I think to make the driver write to stdout, you need the `-o -` flag.

Reply:
Yes, I know it's valid, I'm just not sure it should be allowed. Ideally, all of these temporary files would disappear after the action runs.

Reply:
> Weird. For normal pp, we redirect the action's stdout to the target. With ppx only, we use the `-o` argument. I would stick to using stdout.

did you mean `cppo` here then?

Reply:
No I meant the ppx driver exe. It has a -o option.

By the way, I could rephrase the issue in a much simpler way:

The user action shouldn't introduce its temporary files as additional targets. Ideally, the temporary files would all be deleted after the preprocessing is done.

Probably the simplest thing to do is to introduce a construct for temporary directories where actions can write temp files into. This is still quite a bit more work than I expected though. Let me think about this a bit more.

Reply:
In parallel, I could explore another way to compile the melange repo in a way that doesn't need this feature.